rclcpp/include/rclcpp/node_interfaces/node_graph_interface.hpp

```diff
@@ -182,6 +183,55 @@ class NodeGraphInterface
     const std::string & node_name,
     const std::string & namespace_) const = 0;
 
+  /// Return a map of existing service names and types with a specific node.
+  /**
+   * This function only considers clients - not service servers.
+   * The returned names are the actual names used and do not have remap rules applied.
+   *
+   * \param[in] node_name name of the node
+   * \param[in] namespace_ namespace of the node
+   */
+  RCLCPP_PUBLIC
+  virtual
+  std::map<std::string, std::vector<std::string>>
+  get_client_names_and_types_by_node(
+    const std::string & node_name,
+    const std::string & namespace_) const = 0;
+
+  /// Return a map of existing topic names to list of topic types for a specific node.
```

Comment:
I don't know if one topic name can support multiple types in a node. (Actually, I am afraid not.)
https://github.com/ros2/rmw/blob/ace271650ad39a656d81239050c9639c4190f077/rmw/include/rmw/names_and_types.h#L37

Reply:
thanks, related to https://github.com/ros2/ros2/issues/1095.

Reply:
@ivanpauno 

> but you cannot control what other nodes in the graph do.

Sorry, I don't catch it. Could you give me more information about what other nodes could do in the graph?

all DDS vendor will use  to get names and types from
https://github.com/ros2/rmw_dds_common/blob/9124fc6b915064e171ba24c722be7c9d7fefcecb/rmw_dds_common/src/graph_cache.cpp#L652

no idea why using the [std::set < std::string >](https://github.com/ros2/rmw_dds_common/blob/9124fc6b915064e171ba24c722be7c9d7fefcecb/rmw_dds_common/src/graph_cache.cpp#L637) here, we can see the [demangle_type(item.second.topic_type)](https://github.com/ros2/rmw_dds_common/blob/9124fc6b915064e171ba24c722be7c9d7fefcecb/rmw_dds_common/include/rmw_dds_common/graph_cache.hpp#L345) can only return `std::string`.

It seems we expect one topic name can only have one relative topic type, if so, why do we need to use `vector` there?

Reply:
Sorry for the noise.